framework/aster-frame/src/sync/wait.rs

```diff
@@ -100,80 +103,134 @@ impl WaitQueue {
                 return cond();
             }
 
-            self.enqueue(&waiter);
             waiter.wait();
         }
     }
 
     /// Wake up one waiting thread.
     pub fn wake_one(&self) {
-        while let Some(waiter) = self.waiters.lock_irq_disabled().pop_front() {
+        while let Some(waker) = self.wakers.lock_irq_disabled().pop_front() {
             // Avoid holding lock when calling `wake_up`
-            if waiter.wake_up() {
+            if waker.wake_up() {
                 return;
             }
         }
     }
 
     /// Wake up all waiting threads.
     pub fn wake_all(&self) {
-        while let Some(waiter) = self.waiters.lock_irq_disabled().pop_front() {
+        while let Some(waker) = self.wakers.lock_irq_disabled().pop_front() {
             // Avoid holding lock when calling `wake_up`
-            waiter.wake_up();
+            waker.wake_up();
         }
     }
 
     pub fn is_empty(&self) -> bool {
-        self.waiters.lock_irq_disabled().is_empty()
+        self.wakers.lock_irq_disabled().is_empty()
     }
 
-    // Enqueue a waiter into current waitqueue. If waiter is exclusive, add to the back of waitqueue.
-    // Otherwise, add to the front of waitqueue
-    fn enqueue(&self, waiter: &Arc<Waiter>) {
-        self.waiters.lock_irq_disabled().push_back(waiter.clone());
+    fn enqueue(&self, waker: Arc<Waker>) {
+        self.wakers.lock_irq_disabled().push_back(waker);
     }
 }
 
+/// A waiter that can put the current thread to sleep until it is woken up by the associated
+/// [`Waker`].
+///
+/// By definition, a waiter belongs to the current thread, so it cannot be sent to another thread
+/// and its reference cannot be shared between threads.
 struct Waiter {
-    /// The `Task` held by the waiter.
+    waker: Arc<Waker>,
+}
+
+impl !Send for Waiter {}
+impl !Sync for Waiter {}
+
+/// A waker that can wake up the associated [`Waiter`].
+///
+/// A waker can be created by calling [`Waiter::new`]. This method creates an `Arc<Waker>` that can
+/// be used across different threads.
+struct Waker {
+    has_waken: AtomicBool,
```

Comment:
I'm not clear why we need this `has_waken` flag. In #722, I removed this flag and I think we can directly check task status instead of check the flag. Is there any reason to introduce the flag again?

Reply:
Well, I do think we need this flag.

Note that the waiter may be woken up at time. For example, it may be woken up while running `cond()`. If we remove this flag and encode this flag in the task status, it means that we must set the task state to `TaskStatus::Sleepy` _before_ executing `cond()`. In #722, we update the task status in `wait()`, which means that the wake-up event during `cond()` will be lost.

But it immediately poses another problem: we cannot allow preemption during `cond()`. Otherwise, if a task with `TaskStatus::Sleepy` goes to the background, it will become a sleeping thread permanently and cannot be brought to the foreground until it is explicitly woken up. I do not want to introduce this limitation to `cond()`, so I fall back to the original solution that maintains a `has_waken` flag.

Reply:
> Note that the waiter may be woken up at any time. For example, it may be woken up while running cond()

This only occurs if we enqueue waiter before checking condition. But we can also enqueue waiter after checking condition. For example,

```rust 
impl WaitQueue {
    fn do_wait<F, R>(&self, mut cond: F, timeout: Option<&Duration>) -> Option<R>
    where
        F: FnMut() -> Option<R>,  {
        // ............Code before loop 
        
        loop {
            if let Some(res) = cond() {
                if let Some(timer_callback) = timer_callback {
                    timer_callback.cancel();
                }

                return Some(res);
            };

            if let Some(ref timer_callback) = timer_callback
                && timer_callback.is_expired()
            {
                return cond();
            }

            self.prepear_wait(&waiter);
            waiter.wait();
        }
    }
    
    // Ensure that set task status 
    // and enqueue waiter in atomic context
    fn prepear_wait(&self, waiter: &Arc<Waiter>) {
        let mut waiters = self.waiters.lock_irq_disabled();
        waiter.task.inner_inner_exclusive_access().task_status = TaskStatus::Sleeping;
        waiters.enqueue(waiter.clone);
    }
}
```

Reply:
> This only occurs if we enqueue waiter before checking condition. But we can also enqueue waiter after checking condition.

So how do you deal with the situation where the event occurs just after `cond()` and just before `prepare_wait()`? For example, the network packet arrives exactly at that time, and we are using the `WaitQueue` to wait for a network packet. It's a race condition, isn't it?

Events can really occur at any time, and it's not related to the time when the waiter is enqueued. The event may come:
 - via hardware IRQs that may happen at each instruction as long as the IRQs are not disabled;
 - via soft IRQs that may run concurrently on another CPU core;
 - not to mention the case where kernel preemption is supported and the kernel thread may be preempted just after the `cond()` call.

Reply:
Got it! Thanks for your explanation!